usvm-ts/src/main/kotlin/org/usvm/TSExprTransformer.kt

```diff
@@ -0,0 +1,149 @@
+    }
+
+    fun transform(sort: USort, modifyConstraints: Boolean = true): UExpr<out USort>? = with(ctx) {
+        val (result, type) = when (sort) {
+            fp64Sort -> asFp64() to EtsNumberType
+            boolSort -> asBool() to EtsBooleanType
+            addressSort -> asRef() to null
+            else -> error("")
+        }
+
+        if (modifyConstraints && type != null) suggestType(type)
+
+        return result
+    }
+
+    fun intersectWithTypeCoercion(
+        other: TSExprTransformer,
+        action: CoerceAction
+    ): UExpr<out USort> {
+        intersect(other)
+
+        val rawExprs = exprCache.keys.mapNotNull { sort ->
+            val lhv = transform(sort)
+            val rhv = other.transform(sort)
+            if (lhv != null && rhv != null) {
+                action(lhv, rhv)
+            } else null
+        }
+
+        val innerCoercionExprs = this.generateAdditionalExprs(rawExprs) + other.generateAdditionalExprs(rawExprs)
+
+        val exprs = rawExprs + innerCoercionExprs
+
+        return if (exprs.size > 1) {
+            assert(exprs.all { it.sort == ctx.boolSort })
+            UJoinedBoolExpr(ctx, exprs.cast())
+        } else exprs.single()
+    }
+
+    private fun intersect(other: TSExprTransformer) {
+        exprCache.keys.forEach { sort ->
+            other.transform(sort)
+        }
+        other.exprCache.keys.forEach { sort ->
+            transform(sort)
+        }
+    }
+
+    private val addedExprCache: MutableSet<UExpr<out USort>> = mutableSetOf()
+
+    /**
+     * Generates and caches additional constraints for coercion expression list.
+     *
+     * For now used to save link between fp and bool sorts of [baseExpr].
+     *
+     * @return List of additional [UExpr].
+     */
+    private fun generateAdditionalExprs(rawExprs: List<UExpr<out USort>>): List<UExpr<out USort>> = with(ctx) {
+        if (!rawExprs.all { it.sort == boolSort }) return emptyList()
+        val newExpr = when (baseExpr.sort) {
+            addressSort -> addedExprCache.putOrNull(mkEq(fpToBoolSort(asFp64()), asBool()))
+            else -> null
+        }
+
+        return newExpr?.let { listOf(it) } ?: emptyList()
+    }
+
+    fun asFp64(): UExpr<KFp64Sort> = exprCache.getOrPut(ctx.fp64Sort) {
+        when (baseExpr.sort) {
+            ctx.fp64Sort -> baseExpr
+            ctx.boolSort -> ctx.boolToFpSort(baseExpr.cast())
+            ctx.addressSort -> with(ctx) {
+                TSRefTransformer(ctx, fp64Sort).apply(baseExpr.cast()).cast()
+            }
+
+            else -> ctx.mkFp64(0.0)
+        }
+    }.cast()
+
+
+    fun asBool(): UExpr<KBoolSort> = exprCache.getOrPut(ctx.boolSort) {
+        when (baseExpr.sort) {
+            ctx.boolSort -> baseExpr
+            ctx.fp64Sort -> ctx.fpToBoolSort(baseExpr.cast())
+            ctx.addressSort -> with(ctx) {
+                TSRefTransformer(ctx, boolSort).apply(baseExpr.cast()).cast()
+            }
+
+            else -> ctx.mkFalse()
```

Comment:
Why? If something is not implemented or unexpected it should throw an error

Reply:
Fixed